### PR TITLE
add MS-Celeb-1M pretrained inception_resnet_v1 no-top model support

### PR DESCRIPTION
**Pull request by yunhaia**
add application/inception_resnet_v1 support, and provide pretrained facenet no-top model to download.

### CODE REVIEW

**Comment by fchollet** [comment]
Is there a rationale for supporting InceptionResNetV1 if we already support V2? Why not simply add facenet weights to V2?

**Comment by yunhaia** [comment]
facenet does only provide Inception Resnet V1 pretrained model. That is the reason why need add Inception Resnet v1 model support to keras.

**Comment by fchollet** [comment]
> facenet does only provide Inception Resnet V1 pretrained model. That is the reason why need add Inception Resnet v1 model support to keras.

Could we not simply train the existing InceptionResNetV2 model on the facenet dataset? Maybe it has been done before and we could simply get ahold of the weights?

**Comment by yunhaia** [comment]
> Currently, the best performing model is an Inception-Resnet-v1 model trained on CASIA-Webface aligned with MTCNN.
it seems that the facenet's best performing model is Resnet-v1 per facenet's Readme, and also check facenet's source codes, its train_softmax and train_tripleloss still based on Resnet V1.

**Comment by fchollet** [comment]
> I think it is still necessary to provide Inception Resnet V1 support for keras.

Why? V2 should be better in every way.

> it seems that the facenet's best performing model is Resnet-v1 per facenet's Readme

I assume they haven't tried re-training it since the release of V2.

**Comment by yunhaia** [comment]
> I think it is still necessary to provide Inception Resnet V1 support for keras.
Why? V2 should be better in every way.

In fact, there are no best model from theory, some application models are trained at a concrete version mode. Keras is regarded as an application framework, it should provide several application models for developers to choice and then use.  

> it seems that the facenet's best performing model is Resnet-v1 per facenet's Readme
I assume they haven't tried re-training it since the release of V2.

I think facenet's owner is suitable  to answer this question, and they did not upgrade to use Inception Resnet V2 model may have some reasons, (Resnet v2's performance is not good as Resnet v1 ?? )

**Comment by fchollet** [comment]
Ok, we can add this new model. Could you please find an available reviewer for this PR?

**Comment by yunhaia** [comment]
@myutwo150 can you help to act as a reviewer for this PR since the Keras's Inception Resnet V2 was written by you, and I was inspired to write v1 by your V2 implementation. Thank you. 
@fchollet Thanks.

**Comment by myutwo150** [comment]
Okay. I think I can take a look at it in the next couple of days.

File: keras/applications/inception_resnet_v1.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+Model naming and structure follows TF-slim implementation (which has some additional
+layers and different number of filters from the original arXiv paper):
+https://github.com/davidsandberg/facenet/blob/master/src/models/inception_resnet_v1.py
+
+Pre-trained Facenet weights are 20170512-110547 can be found in:
+https://github.com/davidsandberg/facenet
+
+# Reference
+- [Inception-v4, Inception-ResNet and the Impact of
+   Residual Connections on Learning](https://arxiv.org/abs/1602.07261)
```
Comment (myutwo150) [inline keras/applications/inception_resnet_v1.py]: I think the FaceNet paper should also be added here.

File: keras/applications/inception_resnet_v1.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+                      classes=44051):
+    """Instantiates the Inception-ResNet v1 architecture.
+
+    Optionally loads weights pre-trained on Facenet.
+    Note that when using TensorFlow, for best performance you should
+    set `"image_data_format": "channels_last"` in your Keras config
+    at `~/.keras/keras.json`.
+
+    The model and the weights are compatible with TensorFlow, Theano and
+    CNTK backends. The data format convention used by the model is
```
Comment (myutwo150) [inline keras/applications/inception_resnet_v1.py]: Have you verified that this model gives similar (and reasonable) predictions on all 3 backends?

File: keras/applications/inception_resnet_v1.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        ValueError: in case of invalid argument for `weights`,
+            or invalid input shape.
+    """
+    if weights not in {'facenet', None}:
+        raise ValueError('The `weights` argument should be either '
+                         '`None` (random initialization) or `imagenet` '
+                         '(pre-training on ImageNet).')
+
+    if weights == 'facenet' and include_top and classes != 44051:
+        raise ValueError('If using `weights` as imagenet with `include_top`'
```
Comment (myutwo150) [inline keras/applications/inception_resnet_v1.py]: Not ImageNet.

File: keras/applications/inception_resnet_v1.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+                         ' as true, `classes` should be 44051')
```
Comment (myutwo150) [inline keras/applications/inception_resnet_v1.py]: Why is it 44051?

File: keras/applications/inception_resnet_v1.py [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+                warnings.warn('You are using the TensorFlow backend, yet you '
+                              'are using the Theano '
+                              'image data format convention '
+                              '(`image_data_format="channels_first"`). '
+                              'For best performance, set '
+                              '`image_data_format="channels_last"` in '
+                              'your Keras config '
+                              'at ~/.keras/keras.json.')
+        if include_top:
+            raise ValueError('currently does not support top')
```
Comment (myutwo150) [inline keras/applications/inception_resnet_v1.py]: Why? Also the error message is not correct (the model can be created with `include_top=True, weights=None`) and sounds a bit weird.

File: tests/keras/applications/applications_test.py [hunk 6]
```diff
@@ -174,6 +174,34 @@ def test_inceptionv3_variable_input_channels():
 
 
 @keras_test
+def test_inceptionresnetv1_notop():
```
Comment (myutwo150) [inline tests/keras/applications/applications_test.py]: Please also add the other tests.

File: tests/keras/applications/applications_test.py [hunk 7]
```diff
@@ -174,6 +174,34 @@ def test_inceptionv3_variable_input_channels():
...
+    def target(queue):
+        model = applications.InceptionResNetV1(weights=None, include_top=False)
+        queue.put(model.output_shape)
```
Comment (myutwo150) [inline tests/keras/applications/applications_test.py]: Is the workaround with multiprocessing really necessary for this test? I suppose the Inception-ResNet V1 model should be of similar size to Inception V3. Can you verify it (on CNTK backend)?

**Comment by myutwo150** [comment]
I'm not quite sure if we should name the weights `'facenet'`. I'm not quite familiar to FaceNet, but it seems that the weights are trained in a completely different way to the original FaceNet paper. Is there any good reason to name it `'facenet'`?

**Comment by myutwo150** [comment]
Also please update the docs.

File: keras/applications/inception_resnet_v1.py [hunk 8]
```diff
@@ -0,0 +1,379 @@
+# -*- coding: utf-8 -*-
+"""Facenet Inception-ResNet V1 model for Keras.
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply keras/applications/inception_resnet_v1.py]: ok, will remove Facent

File: keras/applications/inception_resnet_v1.py [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+
+Model naming and structure follows TF-slim implementation (which has some additional
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply keras/applications/inception_resnet_v1.py]: you are right, this line should delete.

File: keras/applications/inception_resnet_v1.py [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+layers and different number of filters from the original arXiv paper):
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply keras/applications/inception_resnet_v1.py]: Yes, same as original paper.

File: keras/applications/inception_resnet_v1.py [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+https://github.com/davidsandberg/facenet/blob/master/src/models/inception_resnet_v1.py
+
+Pre-trained Facenet weights are 20170512-110547 can be found in:
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply keras/applications/inception_resnet_v1.py]: OK

**Comment by yunhaia** [comment]
@myutwo150 very thanks for your review, I will revise it later per your commendation.

File: keras/applications/inception_resnet_v1.py (hunk 4 above)
Comment (yunhaia) [reply keras/applications/inception_resnet_v1.py]: pre-trained FaceNet's classes is 44051.

File: keras/applications/inception_resnet_v1.py [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+
+def InceptionResNetV1(include_top=True,
+                      weights='facenet',
+                      input_tensor=None,
+                      input_shape=None,
+                      pooling=None,
+                      classes=44051):
+    """Instantiates the Inception-ResNet v1 architecture.
+
+    Optionally loads weights pre-trained on Facenet.
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply keras/applications/inception_resnet_v1.py]: you are right, the dataset is MS-Celeb-1M.

File: docs/templates/applications.md [hunk 13]
```diff
@@ -466,6 +467,62 @@ These weights are released under [the Apache License](https://github.com/tensorf
...
+
+A Keras `Model` instance.
+
+### References
+
+- [FaceNet's Inception Resnet V1 imlementation](https://github.com/davidsandberg/facenet/blob/master/src/models/inception_resnet_v1.py)
+
+### License
+
+These weights are released under [the Apache License](https://github.com/tensorflow/models/blob/master/LICENSE).
```
Comment (myutwo150) [inline docs/templates/applications.md]: The license should point to the one in facenet repo.

File: keras/applications/inception_resnet_v1.py [hunk 14]
```diff
@@ -227,13 +222,13 @@ def InceptionResNetV1(include_top=True,
         ValueError: in case of invalid argument for `weights`,
             or invalid input shape.
     """
-    if weights not in {'facenet', None}:
+    if weights not in {'msceleb1m', None}:
         raise ValueError('The `weights` argument should be either '
-                         '`None` (random initialization) or `imagenet` '
-                         '(pre-training on ImageNet).')
+                         '`None` (random initialization) or `msceleb1m` '
+                         '(pre-training on msceleb1m).')
```
Comment (myutwo150) [inline keras/applications/inception_resnet_v1.py]: MS-Celeb-1M.

File: keras/applications/inception_resnet_v1.py [hunk 15]
```diff
@@ -171,15 +166,15 @@ def inception_resnet_block(x, scale, block_type, block_idx, activation='relu'):
...
 
 
-def InceptionResNetV1(include_top=True,
-                      weights='facenet',
+def InceptionResNetV1(include_top=False,
+                      weights='msceleb1m',
                       input_tensor=None,
                       input_shape=None,
                       pooling=None,
                       classes=44051):
```
Comment (myutwo150) [inline keras/applications/inception_resnet_v1.py]: Is 44051 the number of classes in the MS-Celeb-1M dataset?

File: keras/applications/inception_resnet_v1.py [hunk 16]
```diff
@@ -227,13 +222,13 @@ def InceptionResNetV1(include_top=True,
...
+    if weights not in {'msceleb1m', None}:
         raise ValueError('The `weights` argument should be either '
-                         '`None` (random initialization) or `imagenet` '
-                         '(pre-training on ImageNet).')
+                         '`None` (random initialization) or `msceleb1m` '
+                         '(pre-training on msceleb1m).')
 
-    if weights == 'facenet' and include_top and classes != 44051:
-        raise ValueError('If using `weights` as imagenet with `include_top`'
+    if weights == 'msceleb1m' and include_top and classes != 44051:
```
Comment (myutwo150) [inline keras/applications/inception_resnet_v1.py]: It's not necessary to enforce this condition if the weights for the top layers are not provided.

**Comment by myutwo150** [comment]
Please also update the table in applications.md.

**Comment by myutwo150** [comment]
Why can't we support the weights of the top layers? In the facenet repo, it seems that this model is used with an 128-dim bottleneck layer on top of it. I'm not quite sure whether it's really useful to add this model if we do not provide the weights until the bottleneck.

File: docs/templates/applications.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (yunhaia) [reply docs/templates/applications.md]: ok

File: docs/templates/applications.md [hunk 17]
```diff
@@ -466,6 +467,62 @@ These weights are released under [the Apache License](https://github.com/tensorf
...
+    into, only to be specified if `include_top` is `True`, and 
+    if no `weights` argument is specified.
+
+### Returns
+
+A Keras `Model` instance.
+
+### References
+
+- [FaceNet's Inception Resnet V1 imlementation](https://github.com/davidsandberg/facenet/blob/master/src/models/inception_resnet_v1.py)
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply docs/templates/applications.md]: ok

File: docs/templates/applications.md [hunk 18]
```diff
@@ -466,6 +467,62 @@ These weights are released under [the Apache License](https://github.com/tensorf
...
 -----
 
+## InceptionResNetV1
+
+
+```python
+keras.applications.inception_resnet_v1.InceptionResNetV1(include_top=False, weights='msceleb1m', input_tensor=None, input_shape=None, pooling=None, classes=44051)
+```
+
+Inception-ResNet V1 model, with weights pre-trained on MS-Celeb-1M.
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply docs/templates/applications.md]: yes, no ImageNet weights, FaceNet only provides MS-Celeb-1M and CASIA-WebFace weights

File: keras/applications/inception_resnet_v1.py [hunk 19]
```diff
@@ -0,0 +1,374 @@
+# -*- coding: utf-8 -*-
+"""Inception-ResNet V1 model for Keras.
+
+Model naming and structure follows FaceNet's Inception Resnet V1 implementation:
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply keras/applications/inception_resnet_v1.py]: I need check it.

File: keras/applications/inception_resnet_v1.py [hunk 20]
```diff
@@ -1,16 +1,11 @@
...
 https://github.com/davidsandberg/facenet/blob/master/src/models/inception_resnet_v1.py
 
-Pre-trained Facenet weights are 20170512-110547 can be found in:
-https://github.com/davidsandberg/facenet
-
-# Reference
-- [Inception-v4, Inception-ResNet and the Impact of
-   Residual Connections on Learning](https://arxiv.org/abs/1602.07261)
+Pre-trained FaceNet's weights named 20170512-110547 can be downloaded in:
+https://drive.google.com/file/d/0B5MzpY9kBtDVZ2RpVDYwWmxoSUk
```
[the remark this answers is not in the document]
Comment (yunhaia) [reply keras/applications/inception_resnet_v1.py]: ok

**Comment by yunhaia** [comment]
Inception ResNet v1 top model can be support later. No-top Model is very useful in Transfer learning.

**Comment by taehoonlee** [comment]
@yunhaia, Do you have any plans to add the `include_top` model?

**Comment by fchollet** [comment]
Please direct this PR to https://github.com/keras-team/keras-applications (since `applications` is no longer part of this codebase).

**Pull request closed without merging**